src/query/graphite/storage/m3_wrapper.go

```diff
@@ -139,15 +158,23 @@ func translateTimeseries(
 			return nil, errSeriesNoResolution
 		}
 
+		start, end := truncateBoundsToResolution(start, end, resolution)
 		length := int(end.Sub(start) / resolution)
 		millisPerStep := int(resolution / time.Millisecond)
 		values := ts.NewValues(ctx, millisPerStep, length)
 		for _, datapoint := range m3series.Values().Datapoints() {
-			index := int(datapoint.Timestamp.Sub(start) / resolution)
-			if index < 0 || index >= length {
-				// Outside of range requested
+			ts := datapoint.Timestamp
+			if ts.Before(start) {
+				// Outside of range requested.
 				continue
 			}
+
+			if ts.After(end) {
```

Comment:
I think each step is inclusive at the start and exclusive at the end yeah?  i.e. `[start, end)` 

Otherwise a single datapoint would fall into two steps if it aligned with the resolution exactly no?

As such, it sounds like this should be `if !ts.Before(end) { break }`?

Reply:
Also another question... are we worried the values will ever be returned out of order? Heh. Cause if so it'd be better to `continue` here rather than `break`.  It's possible that would be overly defensive programming, but if the underlying storage ever was something else that didn't use our series iterators I suppose that would be possible... (at graphite would be fine since it always uses consolidation and `values.SetValueAt(idx, value)`.

Reply:
Yeah, good point re: `!Before`; was thinking that I may have dropped some points due to truncation so wanted to keep as much as possible, but the dual-read is a good point. 

About the `break` v `continue`, I think that if we have non-ordered series coming in, there's a lot of places that would break worse than this haha. Also given it's `m3_wrapper.go`, I think it's fine to assume it's ordered :P